examples/vision/captcha_ocr.py

```diff
@@ -0,0 +1,502 @@
+        max_length=5,
+        shuffle=True,
+    ):
+        self.data = data
+        self.labels = labels
+        self.char_map = char_map
+        self.batch_size = batch_size
+        self.img_width = img_width
+        self.img_height = img_height
+        self.downsample_factor = downsample_factor
+        self.max_length = max_length
+        self.shuffle = shuffle
+        self.indices = np.arange(len(data))
+        self.on_epoch_end()
+
+    def __len__(self):
+        return int(np.ceil(len(self.data) / self.batch_size))
+
+    def __getitem__(self, idx):
+        # 1. Get the next batch indices
+        curr_batch_idx = self.indices[
+            idx * self.batch_size : (idx + 1) * self.batch_size
+        ]
+        # 2. Only necessay for the last batch
+        batch_len = len(curr_batch_idx)
+        # 3. Instantiate batch arrays
+        batch_images = np.ones(
+            (batch_len, self.img_width, self.img_height, 1), dtype=np.float32
+        )
+        batch_labels = np.ones((batch_len, self.max_length), dtype=np.float32)
+        input_length = np.ones((batch_len, 1), dtype=np.int64) * (
+            self.img_width // self.downsample_factor - 2
+        )
+        label_length = np.zeros((batch_len, 1), dtype=np.int64)
+
+        for j, idx in enumerate(curr_batch_idx):
+            # 4. Get the image and transpose it
+            img = self.data[idx].T
+            # 5. Add extra dimenison
+            img = np.expand_dims(img, axis=-1)
+            # 6. Get the correpsonding label
+            text = self.labels[idx]
+            # 7. Include the pair only if the captcha is valid
+            if is_valid_captcha(text):
+                label = [self.char_map[ch] for ch in text]
+                batch_images[j] = img
+                batch_labels[j] = label
+                label_length[j] = len(text)
+
+        # 8. Make a dictionary of inout data
+        batch_inputs = {
+            "input_data": batch_images,
+            "input_label": batch_labels,
+            "input_length": input_length,
+            "label_length": label_length,
+        }
+        return batch_inputs, np.zeros(batch_len).astype(np.float32)
+
+    def on_epoch_end(self):
+        if self.shuffle:
+            np.random.shuffle(self.indices)
+
+
+# Get a generator object for the training data
+train_data_generator = DataGenerator(
+    data=training_data,
+    labels=training_labels,
+    char_map=char_to_labels,
+    batch_size=batch_size,
+    img_width=img_width,
+    img_height=img_height,
+    downsample_factor=downsample_factor,
+    max_length=max_length,
+    shuffle=True,
+)
+
+# Get a generator object for the validation data
+valid_data_generator = DataGenerator(
+    data=validation_data,
+    labels=validation_labels,
+    char_map=char_to_labels,
+    batch_size=batch_size,
+    img_width=img_width,
+    img_height=img_height,
+    downsample_factor=downsample_factor,
+    max_length=max_length,
+    shuffle=False,
```

Comment:
By line 334 you've downloaded and prepared the data. That's a lot of code for data.

Here's what I recommend:

- make list of image paths and list of their labels
- filter them
- convert them to 2 datasets -- you can use `image_dataset_from_directory` with the `labels` argument, or do it by hand like this: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/preprocessing/image_dataset.py#L209-L233
- map `tf.strings.split` and an `IndexLookup` layer on the labels dataset to convert the strings to integers

This takes at most 20 lines and it will be more performant

Reply:
This functionality is in nightly. Also, the current implementation(the dev branch in nightly) checks if the `labels` argument is an instance of `list` or `tuple`, no checks for `None`. So, if you pass `labels=None`, it will throw error

Reply:
Sorry, I mean `label_mode=None`. You can leave `labels` to its default value.

It's fine to use functionality that's in tf-nightly and experimental. It will soon be in the core API.

Reply:
I have incorporated these changes. Though instead of using `image_from_directory`, I did it manually but it hardly is 20-25 LOC